docs/rfcs/035-physical-replication.md

```diff
@@ -0,0 +1,197 @@
+
+## Impacted components (e.g. pageserver, safekeeper, console, etc)
+
+- Control plane launches the replica
+- Replica Postgres instance connects to the safekeepers, to stream the WAL
+- The primary does not know about the standby, except for the hot standby feedback
+- The primary and replicas all connect to the same pageservers
+
+
+## Decisions, Issues
+
+### Cache invalidation in replica
+
+When a read replica follows the primary in PostgreSQL, it needs to
+stream all the WAL from the primary and apply all the records, to keep
+the local copy of the data consistent with the primary. In Neon, the
+replica can fetch the updated page versions from the pageserver, so
+it's not necessary to apply all the WAL. However, it needs to ensure
+that any pages that are currently in the Postgres buffer cache, or the
+Local File Cache, are either updated, or thrown away so that the next
+read of the page will fetch the latest version.
+
+We choose to apply the WAL records for pages that are already in the
+buffer cache, and skip records for other pages. Somewhat arbitrarily,
+we also apply records affecting catalog relations, fetching the old
+page version from the pageserver if necessary first. See
+neon_redo_read_buffer_filter() function.
+
+The replica wouldn't necessarily need to see all the WAL records, only
+the records that apply to cached pages. For simplicity, we do stream
+all the WAL to the replica, and the replica simply ignores WAL records
+that require no action.
+
+Like in PostgreSQL, the read replica maintains a "replay LSN", which
+is the LSN up to which the replica has received and replayed the
+WAL. The replica can lag behind the primary, if it cannot quite keep
+up with the primary, or if a long-running query conflicts with changes
+that are about to be applied, or even intentionally if the user wishes
+to see delayed data (see recovery_min_apply_delay). It's important
+that the replica sees a consistent view of the whole cluster at the
+replay LSN, when it's lagging behind.
+
+In Neon, the replica connects to a safekeeper to get the WAL
+stream. That means that the safekeepers must be able to regurgitate
+the original WAL as far back as the replay LSN of any running read
+replica. (A static read-only node that does not follow the primary
+does not require a WAL stream however). The primary does not need to
+be running, and when it is, the replicas don't incur any extra
+overhead to the primary (see hot standby feedback though).
+
+### In-progress transactions
+
+In PostgreSQL, when a hot standby server starts up, it cannot
+immediately open up for queries. It first needs to establish a
+complete list of in-progress transactions, including subtransactions,
+that are running at the primary, at the current replay LSN. Normally
+that happens quickly, when the replica sees a "running-xacts" or
+shutdown checkpoint WAL record, because the primary writes a
+running-xacts record at every checkpoint, and in PostgreSQL the
```

Comment:
> because the primary writes a running-xacts record at every checkpoint

How does it work in between checkpoints? Does RW stream running xacts after that as well? Via clog?

Reply:
The running-xact records are written to the WAL. The replica sees them as part of the WAL stream it replays.

Did that answer the question?

Reply:
Yes, but iirc it writes running xacts every 15s or so. If that's right, I still wonder how it works if there are windows when RO doesn't know about running xacts on RW?

Reply:
Ah, now I understand the question. The standby tracks the set of transactions that are running on every replayed record. Whenever it sees a WAL record with an XID that it hasn't seen before, the XID is added to the set of "known assigned XIDs" in memory, and whenever it sees a COMMIT or ABORT record, it removes that XID from the set. That's the set of XIDs that are assumed to be running at that point in time in the primary. The running-xacts record is just needed to bootstrap that process at startup.